qiskit_machine_learning/neural_networks/effective_dimension.py

```diff
@@ -0,0 +1,354 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""An implementation of the effective dimension algorithm."""
+
+import time
+from typing import Optional, Union, List, Callable, Tuple
+
+import numpy as np
+from scipy.special import logsumexp
+
+from qiskit.utils import algorithm_globals
+from qiskit_machine_learning import QiskitMachineLearningError
+
+from .opflow_qnn import OpflowQNN
+from .neural_network import NeuralNetwork
+
+
+class EffectiveDimension:
+    """
+    This class computes the global effective dimension for Qiskit
+    :class:`~qiskit_machine_learning.neural_networks.NeuralNetwork`
+    following the definition used in [1].
+
+        **References**
+        [1]: Abbas et al., The power of quantum neural networks.
+        `The power of QNNs <https://arxiv.org/pdf/2011.00027.pdf>`__.
+    """
+
+    def __init__(
+        self,
+        qnn: NeuralNetwork,
+        weight_samples: Union[List[float], np.ndarray, int] = 1,
+        input_samples: Union[List[float], np.ndarray, int] = 1,
```

Comment:
This I brought up before, but I am not sure its resolved. The type hint here says `List[float]` but should it be List[List[float]] since its really a two dimensional affair i.e. `An array of neural network parameters (weights), of shape               (num_weight_samples, num_weights), or an int`. It could be reshaped from a flat list but I did not see that. 

What I see in the setter for instance is this
```
                len(weight_samples.shape) > 2
                or weight_samples.shape[1] is not self._model.num_weights
```
That checks the dimensions (size of the shape) being > 2 but if 1 will end up failing on the next line as it will index shape[1] which will be out of bounds for a flat list.

I suggested adding unit tests to ensure it ran with plain lists as expected being passed in to make sure of things.

Reply:
While you are absolutely right, I'm not sure I want to keep `List` in type hints at all. They tend to cause many problems, especially when multidimensional arrays are required. I'd prefer to remove support for `List` completely in favor of plain `np.ndarray`. I think I know where this type hint originates from. There's the same problem in `NeuralNetwork`, same list type hint and it is also incorrect, unfortunately. In `NeuralNetwork` a multidimensional array is expected. Usually it is of shape `(num_samples, num_features)`. 

Therefore, `input_samples` may be a single value, a plain list or a list of list at least. I can extend type hints, but prefer to remove support for lists and this is a good chance to start it off. There's an old issue to consider as well: https://github.com/Qiskit/qiskit-machine-learning/issues/159.

Reply:
OK, I removed lists from the type hints.